docs/spec/spec.md

```diff
@@ -0,0 +1,257 @@
+
+## 2. LDAP client
+
+The `ldap:Client` instance needs to be initialized before performing the functionalities. Once initialized, it can perform various operations on directories. Currently, it supports the generic LDAP operations; `add`, `modify`, `modifyDN`, `compare`, `search`, `searchWithType`, `delete`, and `close`.
+
+### 2.1 Configurations
+
+When initializing the `ldap:Client`, `ldap:ConnectionConfig` configuration can be provided.
+
+```ballerina
+# Provides a set of configurations to connect with a directory server.
+#
+# + hostName - The host name of the active directory server
+# + port -  The port of the active directory server
+# + domainName -  The domain name of the active directory
+# + password - The password of the active directory
+public type ConnectionConfig record {|
+    string hostName;
+    int port;
+    string domainName;
+    string password;
+|};
+```
+
+### 2.2 Initialization
+
+The `init` method initializes the `ldap:Client` instance using the parameters `hostName`, `port`, `domainName`, and `password`. The `hostName` and `port` parameters are used to bind the request and authenticate clients with the directory server, while the `domainName` and `password` parameters establish the connection to the server for performing LDAP operations. In case of failure, the method returns an `ldap:Error`."
+
+```ballerina
+# Gets invoked to initialize the LDAP client.
+#
+# + config - The configurations to be used when initializing the client
+# + return - A `ldap:Error` if client initialization failed
+public isolated function init(*ConnectionConfig config) returns Error?;
+```
+
+## 3. Operation types
+
+The currently supported operation types in LDAP are listed here.
+
+### 3.1 Add operation
+
+Creates an entry in a directory server.
+
+```ballerina
+# Creates an entry in a directory server.
+# 
+# + dN - The distinguished name of the entry
+# + entry - The information to add
+# + return - A `ldap:Error` if the operation fails or `ldap:LdapResponse` if successfully created
+remote isolated function add(string dN, Entry entry) returns LdapResponse|Error;
+```
+
+#### 3.1.1 DNs and RDNs
+
+The distinguished name (`DN`) of an entry is used to uniquely identify the entry and its location within the directory information tree (`DIT`) hierarchy. It's similar to how a file path specifies the location of a file in a filesystem.
+
+A `DN` consists of one or more comma-separated components known as relative distinguished names (`RDN`s). Typically, the leftmost component in the `DN` is considered the `RDN` for that entry. [Learn more](https://ldap.com/ldap-dns-and-rdns/)
```

Comment:
These terminologies are currently defined under the `add` operation. but IMO they shouldn’t be, cuz the same terms are used throughout other APIs as well. It would be better to move them out of the add operation and mention them in a separate H3 section. wdyt?

Reply:
Fixed in 5129a1c